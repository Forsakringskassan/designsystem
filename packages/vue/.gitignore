# vite dev
/src/vite-dev/app.vue
