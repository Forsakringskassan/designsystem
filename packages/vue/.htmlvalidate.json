{
    "extends": ["@fkui/vue:recommended"],
    "plugins": ["./htmlvalidate"],
    "transform": {
        "^.*\\.spec.ts$": "html-validate-vue:html"
    }
}
