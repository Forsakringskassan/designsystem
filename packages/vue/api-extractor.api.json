{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/api.d.ts",
    "dtsRollup": {
        "enabled": false
    },
    "messages": {
        "compilerMessageReporting": {
            "default": {
                "logLevel": "none"
            }
        }
    }
}
