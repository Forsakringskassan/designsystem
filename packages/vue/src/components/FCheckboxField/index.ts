export { default as FCheckboxField } from "./FCheckboxField.vue";
export { default as FCheckboxGroupField } from "./FCheckboxField.vue";
export { default as FCheckboxGroup } from "./FCheckboxGroup.vue";
