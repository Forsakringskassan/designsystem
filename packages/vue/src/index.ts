export * from "./components";
export {
    config,
    type FKUIConfig,
    FKUIConfigButtonOrder,
    type FKUIContext,
    type MaybeWithFKUIContext,
    setRunningContext,
} from "./config";
export * from "./internal-components";
export * from "./plugins";
export * from "./types";
export * from "./utils";
