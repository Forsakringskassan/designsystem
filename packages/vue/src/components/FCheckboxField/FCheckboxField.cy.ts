import { type DefineComponent, defineComponent } from "vue";
import { ValidationService } from "@fkui/logic";
import {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "@fkui/test-utils/vue";
import { FFieldsetPageObject } from "../../cypress";
import { FTooltip } from "../FTooltip";
import { FFieldset } from "../FFieldset";
import FCheckboxField from "./FCheckboxField.vue";

function createComponent(template: string): DefineComponent {
    return defineComponent({
        template,
        components: {
            FCheckboxField,
            FFieldset,
            FTooltip,
        },
        mounted() {
            ValidationService.setSubmitted("checkbox-group");
        },
        data() {
            return {
                checkboxModel: [],
            };
        },
    });
}

const defaultTemplate = /* HTML */ `
    <f-fieldset id="checkbox-group" name="checkbox-name" v-validation.required>
        <template #label> Label text </template>
        <template #description="{ descriptionClass }">
            <span :class="descriptionClass"> Description text </span>
        </template>
        <f-checkbox-field v-model="checkboxModel" value="foo">
            Foo
        </f-checkbox-field>
        <f-checkbox-field v-model="checkboxModel" value="bar">
            Bar
        </f-checkbox-field>
    </f-fieldset>
`;

const checkboxField = {
    first: ".checkbox:nth(0)",
    second: ".checkbox:nth(1)",
};

describe("FCheckboxField", () => {
    const fieldset = new FFieldsetPageObject("fieldset");

    it("should have fields with text", () => {
        cy.mount(createComponent(defaultTemplate));

        const checkboxFirst = fieldset.checkBox(checkboxField.first);
        const checkboxSecond = fieldset.checkBox(checkboxField.second);

        fieldset.numberOfCheckboxes().should("be.equal", 2);
        checkboxFirst.el().should("contain.text", "Foo");
        checkboxSecond.el().should("contain.text", "Bar");
    });

    it("should be able to select and deselect fields, expanded should not exist", () => {
        cy.mount(createComponent(defaultTemplate));

        const checkboxFirst = fieldset.checkBox(checkboxField.first);
        const checkboxSecond = fieldset.checkBox(checkboxField.second);

        checkboxFirst.select();
        checkboxSecond.select();
        checkboxFirst.isSelected().should("be.true");
        checkboxSecond.isSelected().should("be.true");

        checkboxFirst.select();
        checkboxSecond.select();
        checkboxFirst.isSelected().should("be.false");
        checkboxSecond.isSelected().should("be.false");
        checkboxFirst.details().should("not.exist");
        checkboxSecond.details().should("not.exist");
    });

    it("should not be able to select disabled fields", () => {
        const template = /* HTML */ `
            <f-fieldset name="checkbox-name" v-validation.required>
                <template #label> Label text </template>
                <f-checkbox-field v-model="checkboxModel" value="foo">
                    Foo
                </f-checkbox-field>
                <f-checkbox-field v-model="checkboxModel" value="bar" disabled>
                    Bar
                </f-checkbox-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));

        const checkboxFirst = fieldset.checkBox(checkboxField.first);
        const checkboxSecond = fieldset.checkBox(checkboxField.second);

        checkboxSecond.select();
        checkboxSecond.isSelected().should("be.false");

        checkboxFirst.select();
        checkboxFirst.isSelected().should("be.true");
    });

    it("should have label with description", () => {
        cy.mount(createComponent(defaultTemplate));

        fieldset.label.el().should("contain.text", "Label text");
        fieldset.label.description().should("contain.text", "Description text");
    });

    it("should not show error message if not required and submitted", () => {
        const template = /* HTML */ `
            <f-fieldset id="checkbox-group" name="checkbox-name">
                <template #label> Label text </template>
                <f-checkbox-field v-model="checkboxModel" value="foo">
                    Foo
                </f-checkbox-field>
                <f-checkbox-field v-model="checkboxModel" value="bar">
                    Bar
                </f-checkbox-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));

        const checkboxFirst = fieldset.checkBox(checkboxField.first);
        checkboxFirst.checkbox().focus().blur();
        fieldset.el().click();
        fieldset.label.errorMessage().should("not.exist");
    });

    it("should show error message when required and submitted", () => {
        cy.mount(createComponent(defaultTemplate));

        fieldset.label.errorMessage().should("exist");
        fieldset.label
            .errorMessage()
            .should("contain.text", "Välj minst ett alternativ.");
    });

    it("should have tooltip", () => {
        const template = /* HTML */ `
            <f-fieldset name="checkbox-name">
                <template #label> Label text </template>
                <template #tooltip>
                    <f-tooltip
                        screen-reader-text="Läs mer om Broschyrer"
                        header-tag="h1"
                    >
                        <template #header> Tooltip header text </template>
                        <template #body> Tooltip body text </template>
                    </f-tooltip>
                </template>
                <f-checkbox-field v-model="checkboxModel" value="foo">
                    Foo
                </f-checkbox-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));

        fieldset.tooltip.iButton().should("exist");
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    it.skip("should always show details regardless of whether foo or bar is checked when show-details is always", () => {
        const template = /* HTML */ `
            <f-fieldset border show-details="always">
                <template #label> Label text </template>
                <template #default>
                    <f-checkbox-field v-model="checkboxModel" value="foo">
                        <template #default> Foo </template>
                        <template #details>
                            Details is always visible
                        </template>
                    </f-checkbox-field>
                    <f-checkbox-field v-model="checkboxModel" value="bar">
                        <template #default> Bar </template>
                        <template #details>
                            Details is always visible
                        </template>
                    </f-checkbox-field>
                </template>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));

        const checkbox1 = fieldset.checkBox(checkboxField.first);
        const checkbox2 = fieldset.checkBox(checkboxField.second);

        checkbox1.select();
        checkbox1.details().should("be.visible");
        checkbox2.details().should("be.visible");
        cy.get("fieldset").should("exist");
        cy.get("fieldset").toMatchScreenshot({ baseDelay: 500 });
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    it.skip("should display details when checkbox is checked and show-details is when-selected", () => {
        const template = /* HTML */ `
            <f-fieldset border show-details="when-selected">
                <template #label> Label text </template>
                <template #default>
                    <f-checkbox-field v-model="checkboxModel" value="foo">
                        <template #default> Foo </template>
                        <template #details>
                            Value foo - Details is visible only when checked
                        </template>
                    </f-checkbox-field>
                    <f-checkbox-field v-model="checkboxModel" value="bar">
                        <template #default> Bar </template>
                        <template #details>
                            Value bar - Details is visible only when checked
                        </template>
                    </f-checkbox-field>
                </template>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));

        const checkbox1 = fieldset.checkBox(checkboxField.first);
        const checkbox2 = fieldset.checkBox(checkboxField.second);

        checkbox1.select();
        checkbox1.details().should("be.visible");
        checkbox2.details().should("not.exist");
        checkbox1.select();
        checkbox2.select();
        checkbox1.details().should("not.exist");
        checkbox2.details().should("be.visible");
        cy.get("fieldset").should("exist");
        cy.get("fieldset").toMatchScreenshot({ baseDelay: 500 });
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    describe.skip("density", () => {
        const DensityComponent = defineComponent({
            template: /* HTML */ `
                <density-wrapper>
                    <f-fieldset v-validation.required>
                        <template #label>
                            <h2>Kryssrutegrupp</h2>
                        </template>
                        <f-checkbox-field
                            v-model="checkboxField"
                            value="Kryssruta1"
                        >
                            Kryssruta
                        </f-checkbox-field>
                        <f-checkbox-field
                            v-model="checkboxField"
                            value="Kryssruta2"
                        >
                            Kryssruta
                        </f-checkbox-field>
                        <f-checkbox-field
                            v-model="checkboxField"
                            value="Kryssruta3"
                        >
                            Kryssruta
                        </f-checkbox-field>
                        <f-checkbox-field
                            v-model="checkboxField"
                            value="Kryssruta4"
                        >
                            Kryssruta
                        </f-checkbox-field>
                    </f-fieldset>
                </density-wrapper>
            `,
            components: {
                DensityWrapper,
                FFieldset,
                FCheckboxField,
            },
        });

        it(`should be densified`, () => {
            cy.viewport(densityWrapperWidth, densityWrapperHeight);
            cy.mount(DensityComponent);
            cy.toMatchScreenshot();
        });
    });
});
