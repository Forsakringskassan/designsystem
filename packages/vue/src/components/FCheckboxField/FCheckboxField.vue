<script lang="ts">
import { defineComponent, toValue } from "vue";
import { ElementIdService, type ValidityEvent } from "@fkui/logic";
import isEqual from "lodash/isEqual";
import { dispatchComponentValidityEvent, renderSlotText, hasSlot, getHTMLElementFromVueRef } from "../../utils";
import { useFieldset } from "../FFieldset";

// ! Boolean must be last so that empty string ('') is not parsed as true, see: https://github.com/vuejs/vue/blob/81e1e47cabbd479e2a285f03120944f1efffe749/test/unit/features/options/props.spec.js#L516.
const anyType = [String, Object, Array, Number, Date, Boolean];

export default defineComponent({
    name: "FCheckboxField",
    inheritAttrs: false,
    props: {
        /**
         * Set to `true`, empty string `""` or string `"disabled"` to disable this input field.
         */
        disabled: {
            type: Boolean,
            required: false,
            default: false,
        },
        /**
         * The id for the input id attribute.
         * The id for the label for attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * The value for the input checked attribute.
         * @model
         */
        // ? The rule is disabled so that the `checked` prop can be undefined or null.
        /* eslint-disable-next-line vue/require-default-prop -- technical debt,
        /* it should contain a default value of undefined and proptype should
        /* include undefined (see comment on line above) */
        modelValue: {
            type: anyType,
            required: false,
        },
        /**
         * The value for the input.
         */
        value: {
            type: anyType,
            required: true,
        },
    },
    emits: ["change", "update:modelValue"],
    setup() {
        const { showDetails, getFieldsetLabelText } = useFieldset();
        return {
            showDetails,
            getFieldsetLabelText,
        };
    },
    data() {
        return {
            expanded: false,
            height: 0,
            initialStyle: {
                overflow: "hidden",
                transition: "height 400ms cubic-bezier(0.46, 0.03, 0.52, 0.96)",
            },
            hiddenStyle: {
                height: "auto",
                position: "absolute",
                visibility: "hidden",
            },
            visibleStyle: {
                width: "",
                position: "",
                visibility: "",
                height: "0px",
            },
            openedStyle: {
                height: "auto",
            },
        };
    },
    computed: {
        attrs(): Record<string, unknown> {
            let checked;
            if (Array.isArray(this.modelValue)) {
                // `isEqual` since can't compare equality with a reactive (modelValue) object and non-reactive (this.value) object.
                checked = this.modelValue.findIndex((it) => isEqual(toValue(it), toValue(this.value))) >= 0;
            } else {
                checked = this.value === this.modelValue;
            }

            return {
                ...this.$attrs,
                value: this.value,
                checked,
                onChange: (event: Event) => {
                    if (event.target instanceof HTMLInputElement) {
                        this.emitVModelEvent(event);
                    }
                },
                onInput: (event: Event) => {
                    // ? Force focus on Mac + Safari
                    (event.target as HTMLInputElement).focus();
                },
            };
        },
        disabledClass(): string {
            return this.disabled ? "disabled" : "";
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any -- technical debt, nore sure whats going on here
        injected(): any {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- technical debt, nore sure whats going on here
            return this as any;
        },
    },
    methods: {
        updateExpandedFlag(): void {
            const checkboxInput = getHTMLElementFromVueRef(this.$refs["checkboxInput"]) as HTMLInputElement;
            this.expanded = checkboxInput.checked;
        },
        emitVModelEvent(event: Event): void {
            let newModel;

            if (Array.isArray(this.modelValue)) {
                // `isEqual` since can't compare equality with a reactive (modelValue) object and non-reactive (this.value) object.
                newModel = [...this.modelValue].filter((it) => !isEqual(toValue(it), toValue(this.value)));

                // If nothing was filtered out (unchecked), then it was instead checked and the value should be pushed to model.
                if (this.modelValue.length <= newModel.length) {
                    newModel.push(this.value);
                }
            } else {
                if (this.value === this.modelValue) {
                    newModel = typeof this.value === "boolean" ? false : undefined;
                } else {
                    const target = event.target as HTMLInputElement;
                    newModel = target.value === "true" ? true : this.value;
                }
            }

            /**
             * V-model event.
             * @event update:modelValue
             * @type {anyType | anyType[]}
             */
            this.$emit("update:modelValue", newModel);

            /**
             * Emitted when the value of the checkbox changes.
             *
             * @event change
             * @type {anyType | anyType[]}
             */
            this.$emit("change", newModel);
        },
        onKeydown(event: Event): void {
            // Fix SFKUI-2912: mark checkbox with space in IE11
            event.stopPropagation();
        },
        onValidity({ detail }: CustomEvent<ValidityEvent>): void {
            if (detail.target !== this.$el.querySelector("input")) {
                return;
            }

            let errorMessage = "";

            if (hasSlot(this, "default")) {
                const labelText = this.injected.getFieldsetLabelText();
                if (labelText) {
                    errorMessage = `${labelText} ${renderSlotText(this.$slots.default)}`;
                } else {
                    errorMessage = `${renderSlotText(this.$slots.default)}`;
                }
            }

            const element = this.$el.querySelector(`#${detail.elementId}`);
            if (element) {
                dispatchComponentValidityEvent(element, {
                    ...detail,
                    errorMessage,
                    focusElementId: detail.elementId,
                });
            }
        },
        enter(element: Element): void {
            const htmlElement = getHTMLElementFromVueRef(element);
            const computedStyle = getComputedStyle(element);
            Object.assign(htmlElement.style, this.initialStyle);
            Object.assign(htmlElement.style, this.hiddenStyle);
            htmlElement.style.width = computedStyle.width;
            const height = computedStyle.height;
            Object.assign(htmlElement.style, this.visibleStyle);
            // Force redraw
            /* eslint-disable-next-line @typescript-eslint/no-unused-expressions -- technical debt, there should be a better way */
            getComputedStyle(element).height;
            setTimeout(() => {
                this.height = parseInt(height as string, 10);
                htmlElement.style.height = height;
            });
        },
        afterEnter(element: Element): void {
            const htmlElement = getHTMLElementFromVueRef(element);
            Object.assign(htmlElement.style, this.openedStyle);
        },
        leave(element: Element): void {
            const htmlElement = getHTMLElementFromVueRef(element);
            const height = getComputedStyle(element).height;
            htmlElement.style.height = height;

            // Force redraw
            /* eslint-disable-next-line @typescript-eslint/no-unused-expressions -- technical debt, there should be a better way */
            getComputedStyle(element).height;
            setTimeout(() => {
                Object.assign(htmlElement.style, this.visibleStyle);
            });
        },
    },
});
</script>

<template>
    <div class="checkbox" :class="disabledClass" @validity="onValidity">
        <input
            :id="id"
            v-bind="attrs"
            ref="checkboxInput"
            type="checkbox"
            class="checkbox__input"
            :disabled="disabled"
            @keydown.space="onKeydown"
            @change="updateExpandedFlag()"
        />
        <label :class="$slots.details ? 'checkbox__label checkbox__width' : 'checkbox__label'" :for="id">
            <!-- @slot Slot for label content. -->
            <slot name="default"></slot>
            <template v-if="$slots.details">
                <span v-if="showDetails === 'always'" class="checkbox__details">
                    <br />
                    <!-- @slot Slot for extended label, should only contain short text -->
                    <slot name="details"></slot>
                </span>
                <transition
                    v-if="showDetails === 'when-selected'"
                    @enter="enter"
                    @after-enter="afterEnter"
                    @leave="leave"
                >
                    <span v-if="expanded" class="checkbox__details">
                        <br />
                        <!--
                        @slot Slot for details, should only contain short text
                        @binding {number} height The height of the expanded details content.
                        -->
                        <slot name="details" :height="height"></slot>
                    </span>
                </transition>
            </template>
        </label>
    </div>
</template>
