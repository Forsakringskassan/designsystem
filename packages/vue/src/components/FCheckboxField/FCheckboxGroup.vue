<template>
    <f-fieldset :id="id" :name="name" v-bind="$attrs">
        <template #label>
            <!-- @slot Slot for label content. This slot is required. -->
            <slot name="label"></slot>
        </template>

        <template v-if="$slots.tooltip" #tooltip>
            <!-- @slot Slot for tooltip. -->
            <slot name="tooltip"></slot>
        </template>

        <template #description="{ descriptionClass, discreteDescriptionClass }">
            <!--
                @slot Optional slot for description. See [FLabel](#/Components/FLabel) for details.
                @binding {string[]} descriptionClass CSS classes for primary description content.
                @binding {string[]} discreteDescriptionClass CSS classes for format description.
            -->
            <slot name="description" v-bind="{ descriptionClass, discreteDescriptionClass }"></slot>
        </template>

        <template #error-message>
            <!-- @slot Slot for displaying single or several error messages. -->
            <slot name="error-message"></slot>
        </template>

        <!-- @slot Slot for main content, usually checkboxes. -->
        <slot name="default"></slot>
    </f-fieldset>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService } from "@fkui/logic";
import { FFieldset } from "../FFieldset";
import { TranslationMixin } from "../../plugins";

/**
 * @deprecated `<f-checkbox-group>` is deprecated.
 * See migration guide: https://forsakringskassan.github.io/latest/guide/migration/migrating-to-fieldset.html
 */
export default defineComponent({
    name: "FCheckboxGroup",
    components: { FFieldset },
    mixins: [TranslationMixin],
    inheritAttrs: false,
    props: {
        /**
         * The id for the fieldset id attribute.
         * If the prop is not set the id will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * The name of the checkbox group.
         */
        name: {
            type: String,
            required: true,
        },
    },
});
</script>
