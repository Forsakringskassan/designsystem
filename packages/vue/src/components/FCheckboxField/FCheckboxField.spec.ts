import { VueWrapper, mount } from "@vue/test-utils";
import { createPlaceholderInDocument } from "@fkui/test-utils/vue";
import FCheckboxField from "./FCheckboxField.vue";

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FCheckboxField, {
        attrs: { ...attrs },
        props: { value: "Default value", ...props },
        slots: { default: "Default label", ...slots },
        attachTo: createPlaceholderInDocument(),
    });
}

it.each`
    vModel       | value    | expected
    ${undefined} | ${false} | ${false}
    ${null}      | ${false} | ${false}
    ${""}        | ${false} | ${false}
    ${false}     | ${""}    | ${false}
    ${0}         | ${false} | ${false}
    ${undefined} | ${true}  | ${false}
    ${null}      | ${true}  | ${false}
    ${""}        | ${true}  | ${false}
    ${true}      | ${""}    | ${false}
    ${0}         | ${true}  | ${false}
    ${0}         | ${0}     | ${true}
    ${1}         | ${0}     | ${false}
    ${1}         | ${1}     | ${true}
    ${0}         | ${1}     | ${false}
`(
    'should handle v-model value "$vModel" where checked should be "$expected" when checkbox value is "$value"',
    ({ vModel, value, expected }) => {
        const wrapper = createWrapper({
            props: {
                value,
                modelValue: vModel,
            },
        });

        expect(wrapper.get("input").element.checked).toBe(expected);
    },
);

describe("snapshots", () => {
    it("should match snapshot with label and input", () => {
        const wrapper = createWrapper();
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("attributes", () => {
    it("should pass attributes", () => {
        const wrapper = createWrapper({
            attrs: {
                disabled: true,
                required: true,
            },
        });
        const input = wrapper.get("input");

        expect(input.attributes("disabled")).toBeDefined();
        expect(input.attributes("required")).toBeDefined();
    });
});

describe("disabled", () => {
    it.each`
        disabled      | expectedResult | description
        ${"disabled"} | ${true}        | ${"be"}
        ${true}       | ${true}        | ${"be"}
        ${false}      | ${false}       | ${"not be"}
    `(
        "should $description disabled when disabled prop is $disabledAttribute",
        async ({ disabled, expectedResult }) => {
            const wrapper = createWrapper({
                props: {
                    disabled,
                },
            });
            const input = wrapper.get("input").element;
            expect(input.disabled).toBe(expectedResult);
            expect(wrapper.classes("disabled")).toBe(expectedResult);
        },
    );
});

describe("events", () => {
    it("should support v-model by emitting update:modelValue event with value", async () => {
        const wrapper = createWrapper({
            props: { value: "Some value", modelValue: "Some value" },
        });

        const input = wrapper.get("input");
        const htmlInput = input.element;

        expect(htmlInput.checked).toBe(true);
        await wrapper.setProps({ modelValue: undefined });
        expect(htmlInput.checked).toBe(false);

        await input.trigger("click");
        expect(
            wrapper.emitted("update:modelValue")![0][0],
        ).toMatchInlineSnapshot(`"Some value"`);
    });

    it("should emit change event when input value changes", async () => {
        const wrapper = createWrapper({
            props: { value: true, modelValue: false },
        });

        const input = wrapper.get("input");

        await input.trigger("click");
        expect(wrapper.emitted("change")![0]).toMatchInlineSnapshot(`
            [
              true,
            ]
        `);
    });

    describe("should support v-model as array", () => {
        it("should add value to array", async () => {
            const wrapper = createWrapper({
                props: {
                    value: "This checkbox",
                    modelValue: ["Another checkbox"],
                },
            });

            await wrapper.get("input").trigger("click");
            expect(wrapper.emitted("update:modelValue")![0][0])
                .toMatchInlineSnapshot(`
                [
                  "Another checkbox",
                  "This checkbox",
                ]
            `);
        });

        it("should remove value from array", async () => {
            const wrapper = createWrapper({
                props: {
                    value: "This checkbox",
                    modelValue: ["Another checkbox", "This checkbox"],
                },
            });

            await wrapper.get("input").trigger("click");
            expect(wrapper.emitted("update:modelValue")![0][0])
                .toMatchInlineSnapshot(`
                [
                  "Another checkbox",
                ]
            `);
        });

        it("should add nested array", async () => {
            const wrapper = createWrapper({
                props: {
                    value: ["This checkbox"],
                    modelValue: ["Another checkbox"],
                },
            });

            await wrapper.get("input").trigger("click");
            expect(wrapper.emitted("update:modelValue")![0][0])
                .toMatchInlineSnapshot(`
                [
                  "Another checkbox",
                  [
                    "This checkbox",
                  ],
                ]
            `);
        });

        it("should remove nested array", async () => {
            const wrapper = createWrapper({
                props: {
                    value: ["This checkbox"],
                    modelValue: ["Another checkbox", ["This checkbox"]],
                },
            });

            await wrapper.get("input").trigger("click");
            expect(wrapper.emitted("update:modelValue")![0][0])
                .toMatchInlineSnapshot(`
                [
                  "Another checkbox",
                ]
            `);
        });

        it("should add nested object", async () => {
            const wrapper = createWrapper({
                props: {
                    value: { foo: true },
                    modelValue: ["Another checkbox"],
                },
            });

            await wrapper.get("input").trigger("click");
            expect(wrapper.emitted("update:modelValue")![0][0])
                .toMatchInlineSnapshot(`
                [
                  "Another checkbox",
                  {
                    "foo": true,
                  },
                ]
            `);
        });

        it("should remove nested object", async () => {
            const wrapper = createWrapper({
                props: {
                    value: { foo: true },
                    modelValue: ["Another checkbox", { foo: true }],
                },
            });

            await wrapper.get("input").trigger("click");
            expect(wrapper.emitted("update:modelValue")![0][0])
                .toMatchInlineSnapshot(`
                [
                  "Another checkbox",
                ]
            `);
        });
    });

    it("should pass listeners", async () => {
        const focus = jest.fn();
        const blur = jest.fn();

        const wrapper = createWrapper({
            attrs: {
                onFocus: focus,
                onBlur: blur,
            },
        });
        const input = wrapper.get("input");
        await input.trigger("focus");
        await input.trigger("blur");

        expect(focus).toHaveBeenCalled();
        expect(blur).toHaveBeenCalled();
    });

    it("should pass click listener and trigger focus on click", async () => {
        const click = jest.fn();

        const wrapper = createWrapper({
            attrs: { onClick: click },
        });

        const input = wrapper.get("input");
        const htmlInput = input.element;
        htmlInput.focus = jest.fn();

        await input.trigger("click");
        expect(click).toHaveBeenCalled();
        expect(htmlInput.focus).toHaveBeenCalled();
    });
});
