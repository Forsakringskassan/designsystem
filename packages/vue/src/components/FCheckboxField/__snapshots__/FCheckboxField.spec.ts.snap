// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot with label and input 1`] = `
<div
  class="checkbox"
>
  <input
    class="checkbox__input"
    id="fkui-vue-element-0001"
    type="checkbox"
    value="Default value"
  />
  <label
    class="checkbox__label"
    for="fkui-vue-element-0001"
  >
    
    Default label
    
    <!--v-if-->
  </label>
</div>
`;
